recovery/elliptics_recovery/types/merge.py

```diff
@@ -733,114 +270,153 @@ def recover(self, keys):
         '''
         log.info("Server-send bucket: num keys: {0}".format(len(keys)))
 
-        def contain(key, ranges):
-            index = bisect(ranges, key)
-            return index % 2 == 1
+        def contain(address, backend_id, key):
+            addr, _, backend = self.routes.get_id_routes(key)[0]
+            return address == addr and backend_id == backend
 
-        responses = dict([(str(k), []) for k in keys]) # key -> [list of responses]
-        for addr, backend_id, backend_ranges in self.routes:
-            key_candidates = [k for k in keys if not contain(k, backend_ranges)]
-            if key_candidates:
-                self._server_send(key_candidates, addr, backend_id, responses)
+        responses = {str(k): [] for k in keys} # key -> [list of responses]
+        for addr, backend_id in self.backends:
+            key_candidates = [k for k in keys if not contain(addr, backend_id, k)]
+            for i in range(self.ctx.attempts):
+                if key_candidates:
+                    timeouted_keys = self._server_send(key_candidates, addr, backend_id, responses)
+                    key_candidates = timeouted_keys
+            self._update_timeouted_keys_stats(len(key_candidates))
 
         self._remove_bad_keys(responses)
-        return self._get_unrecovered_keys(responses)
+
+        return not self._has_unrecovered_keys(responses)
 
     def _server_send(self, keys, addr, backend_id, responses):
         '''
         Calls server-send with a given list of keys to the specific backend.
+        Returns list of timeouted keys.
         '''
         log.debug("Server-send: address: {0}, backend: {1}, num keys: {2}".format(addr, backend_id, len(keys)))
 
+        if self.ctx.dry_run:
+            return []
+
+        start_time = time.time()
+        recovers_in_progress = len(keys)
+
         self.session.set_direct_id(addr, backend_id)
-        iterator = self.session.server_send(keys, elliptics.iterator_flags.move, list(self.session.groups))
-        for result in iterator:
+        flags = 0 if self.ctx.safe else elliptics.iterator_flags.move
+        iterator = self.session.server_send(keys, flags, list(self.session.groups))
+
+        timeouted_keys = []
+        index = -1
+        for index, result in enumerate(iterator):
             status = result.response.status
+            self._update_stats(start_time, index + 1, recovers_in_progress, status)
```

Comment:
enumerate accepts optional parametr, which sets initial value of index. It's 0 by default.
```python 
enumerate(iterator, 1)
```
 to remove `index + 1`